keyutils-raw/src/types.rs

```diff
@@ -0,0 +1,86 @@
+// WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+// ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+// (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+// LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+// ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+/// Alias for the key_serial_t kernel type, representing a keyring (or key).
+pub type KeyringSerial = std::num::NonZeroI32;
+
+/// Alias for the key_perm_t kernel type, representing a keyring's (or key's)
+/// permission bits.
+///
+/// See `Permission`.
+pub type KeyPermissions = u32;
+
+pub type TimeoutSeconds = libc::c_uint;
+
+/// An enumeration for the keyrings which may be set as the default.
+///
+/// Keys which are implicitly required via syscalls and other operations are
+/// placed in the default keyring.
+#[derive(Debug, PartialEq, Eq)]
+pub enum DefaultKeyring {
+    /// Do not change the default keyring.
+    ///
+    /// This may be used to get the current default keyring.
+    NoChange = -1,
+    /// Set the thread-specific keyring as the default.
+    ThreadKeyring = 1,
+    /// Set the process-specific keyring as the default.
+    ProcessKeyring = 2,
+    /// Set the session-specific keyring as the default.
+    SessionKeyring = 3,
+    /// Set the user-specific keyring as the default.
+    UserKeyring = 4,
+    /// Set the user session-specific keyring as the default.
+    UserSessionKeyring = 5,
+    /// Set the user session-specific keyring as the default.
+    GroupKeyring = 6,
+    /// Set the default keyring to the default logic.
+    ///
+    /// Keys will be placed in the first available keyring of:
+    ///
+    ///   - thread-specific
+    ///   - process-specific
+    ///   - session-specific
+    ///   - user-specific
+    DefaultKeyring = 0,
+}
+
+impl From<libc::c_long> for DefaultKeyring {
+    fn from(id: libc::c_long) -> DefaultKeyring {
+        use self::DefaultKeyring::*;
+        match id {
+            x if x == NoChange as libc::c_long => NoChange,
+            x if x == ThreadKeyring as libc::c_long => ThreadKeyring,
+            x if x == ProcessKeyring as libc::c_long => ProcessKeyring,
+            x if x == SessionKeyring as libc::c_long => SessionKeyring,
+            x if x == UserKeyring as libc::c_long => UserKeyring,
+            x if x == UserSessionKeyring as libc::c_long => UserSessionKeyring,
+            x if x == GroupKeyring as libc::c_long => GroupKeyring,
+            x if x == DefaultKeyring as libc::c_long => DefaultKeyring,
+            _ => panic!("Invalid value for a default keyring: {}", id),
```

Comment:
I agree, I kept the From impl as that what we were doing before. TryFrom initially seems like better semantics, but there’s a question of what to do when “keyctl_set_reqkey_keyring” returns a variant we don’t understand. Having the function return an error seems wrong, as the call succeed.

I think there are two reasonable options here:
  1. Return a “Result<Option<DefaultKeyring>>”
  2. Return a “Result<DefaultKeyring>” and have an UNKNOWN variant in DefaultKeyring.

I’ll implement Option 1, if that sounds reasonable to you.

Reply:
Hmm. We have `Other` variants in some of the key payloads types. Popping a `warning!` or `debug!` in when that branch gets hit with a pointer to file an issue would be useful too. This would allow callers to restore the default back within a scope as well.

Reply:
So I changed this to have keyutils-raw's `DefaultKeyring` implement `TryFrom`, and have the keyutils's function `set_default` just do `ret.try_into().unwrap()`. This allows the function signature to remain the same. We should review this use of `unwrap()` along with all the others in a separate PR addressing #30